### DISCUSSION TITLE
Field of type `pd.Timestamp` needs to be added with `ConfigDict(arbitrary_types_allowed=True)` in `pydantic` v2

### DISCUSSION THREAD

**Discussion by andresliszt** [pydantic/pydantic#6972] [Pydantic V2]
pydantic < 2 supports `pd.Timestamp` as a valid field type, the example below is valid for pydantic < 2

```python

import pandas as pd
from pydantic import BaseModel


class Model(BaseModel):
    created_ts: pd.Timestamp

if __name__ == "__main__":
    # passing it as a valid timestamp string
    model = Model(created_ts="2023-08-01 00:00:00")
    print(model.created_ts)
    #> 2023-08-01 00:00:00
   # passing as a pandas Timestamp and also adding timezone
    model = Model(created_ts=pd.Timestamp("2023-08-01 00:00:00", tz = "UTC"))
    print(model.created_ts)
    #> 2023-08-01 00:00:00+00:00

```

In pydantic v2 this is not allowed anymore

```python

import pandas as pd
from pydantic import BaseModel


class Model(BaseModel):
    created_ts: pd.Timestamp

if __name__ == "__main__":
    model = Model(created_ts=pd.Timestamp("2023-08-01 00:00:00", tz = "UTC"))
    print(model.created_ts)
    #> pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 
    #> 'pandas._libs.tslibs.timestamps.Timestamp'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error  
    #> or implement `__get_pydantic_core_schema__` on your type to fully support it.

```

So i said, what about if i add `model_config = ConfigDict(arbitrary_types_allowed=True)`, now it accepts `pd.Timestamp` as field but it doesn't cast a valid timestamp string into a `pd.Timestamp` as the first example. To cast it, a solution could be add a `field_validator` to cast valid string into `pd.Timestamp` typos, but when you have **several** timestamp fields (as my personal case) it could be really annoying. 

So my question here, this is deprecated? Should i start considering declare those fields as `datetime.datetime`, what about timezones?

**EDIT**: This is more generic, `pd.Timestamp` is a subclass of `datetime.datetime`, the last is allowed by pydantic. Now pydantic doesn't recognize subclass fields

**Comment by andresliszt** [comment]
related with #6979

**Comment by s-celles** [comment]
Could you show with an example how to handle that type ?

**Comment by gusutabopb** [comment]
@scls19fr Here's how I handled it:

```python
import datetime
from pydantic import PlainValidator, WithJsonSchema, BaseModel
from typing import Annotated

_Timestamp = Annotated[
    pd.Timestamp,
    PlainValidator(lambda x: pd.Timestamp(x)),
    WithJsonSchema({"type": 'date-time'})
]

class MyModel(BaseModel):
    ts: _Timestamp
    n: int
        
model = MyModel(ts="2023-08-01", n=10)

assert isinstance(model.ts, datetime.datetime)
assert isinstance(model.ts, pd.Timestamp)

print(model.model_dump())
print(model.model_dump_json())
print(MyModel.model_json_schema())
```

Output

```python
{'ts': Timestamp('2023-08-01 00:00:00'), 'n': 10}
{"ts":"2023-08-01T00:00:00","n":10}
{'properties': {'ts': {'title': 'Ts', 'type': 'date-time'}, 'n': {'title': 'N', 'type': 'integer'}}, 'required': ['ts', 'n'], 'title': 'MyModel', 'type': 'object'}
```



For some reason using `PlainValidator(pd.Timestamp)` did not work, thus the seeminlgy superflous lambda (`PlainValidator(lambda x: pd.Timestamp(x))`.

`WithJsonSchema({"type": 'date-time'})` is not needed for JSON dumping, but is needed to generate the JSON schema. Removing it makes the `.model_json_schema()` call throw `PydanticInvalidForJsonSchema`.


I'm not sure if this is the best approach, but it's what I managed to get to work with my limited Pydantic v2 experience. Happy to hear about alternative/better approaches.

**Reply by mazzma12** [reply]
hey, any way to make it run without the `from typing import Annotated` and using the new pydantic syntax using  `__get_pydantic_core_schema__` ?

**Reply by gusutabopb** [reply]
What's the issue with the `Annotated` syntax? Under the hood, `__get_pydantic_core_schema__` is used by both `PlainValidator` and `WithJsonSchema`. IMHO implementing `__get_pydantic_core_schema__` directly is way more verbose/low-level. Pydantic itself implements many of its types using `Annotated` (see [types.py](https://github.com/pydantic/pydantic/blob/main/pydantic/types.py#L234)).

Not sure if this is an optimal approach, but this worked for me:

```python
from typing import Any
from pydantic_core import CoreSchema, core_schema
from pydantic import GetCoreSchemaHandler, GetJsonSchemaHandler

class _Timestamp(pd.Timestamp):
    
    @classmethod
    def __get_pydantic_json_schema__(cls, core_schema: CoreSchema, handler: GetJsonSchemaHandler):
        return {'type': 'date-time'}

    @classmethod
    def __get_pydantic_core_schema__(cls, source_type: Any, handler: GetCoreSchemaHandler):
        return core_schema.no_info_plain_validator_function(pd.Timestamp)
```

For more advanced cases you'd probably need to dive deep into pydantic/pydantic-core internals to understand what you're doing. I will stick with `Annotated` whenever possible.

**Comment by mazzma12** [comment]
It was to run on a legacy project with python 3.8 . Thank you it's working as expected :)
